ui/flow_of_allowing_to_add_models.md

```diff
@@ -0,0 +1,123 @@
+If you want to also add a "new" button inside you textual summary page you need to do modification to `plugins/manageiq-ui-classic/app/heplers/application_helpers/toolbar/your_item_center.rb`. But this is not part of this guide.
+
+#### Add button class
+Create a button class to handle "new" operation. 
+It needs to be in `manageiq-ui-classic/app/helpers/application_helper/button/your_item_new.rb`. You can copy another button and adjust it.
+
+## JS Form
+You need to create a form-folder in `manageiq-ui-classic/app/javascript/components/your-item-form/`. With `index.jsx` and `your-item-form.schema.js`.
+You can copy from another form and adjust as needed.
+
+Here you determine the GUI form the user will see when creating a new instance of your item. I won't go into the details of it in this guide.\
+When developing, it's safest to start with a minial form just to see that it's working.
+
+Note that the fields on your form will constitute the `options` argument for your `self.raw_create_your_item` method.
+
+After you add your form, you need to register it in `component-definitions-common.js`.
+
+## The UI Controller
+In your controller in `plugins/manageiq-ui-classic/app/controllers/your_item_controller.rb` you need to add the `new` method.\
+You can copy it from another controller and adjust as needed.
+
+Another thing you need in your controller is a method called `specific_buttons(pressed)`. If you don't have it already, copy it from another controller.\
+The method consists of a `case`.\
+Add the following `when` to the case. If you have just copied the method from another controller you can remove all other `when`s.
+```ruby
+when 'your_item_new'
+  javascript_redirect(:action => 'new')
+end
+```
+
+## haml template for 'new' operation
+Create `new.html.haml` under `plugins/manageiq-ui-classic/app/views/your_item/`. You can copy it from another item and adjust as needed.
+Make sure to point at your JS Form which you defined above `react 'YourItemForm'`
+
+## UI routes
+In `manageiq-ui-classic/config/routes.rb`, you need to find the section for your_item. Then you need to make sure there's `new` in the get section, and `button` in the post section.
+
+## API Controller
+If there isn't one, create a controller for your item in `manageiq-api/app/controllers/api/your_items_controller.rb`. You can copy one of the existing controllers and adjust to your item.
+
+Make sure that there is a `create_resource` method in the api controller. If not, copy if from some other controller and adjust to fit your item.\
+This method is supposed to call the method`create_your_item_queue` on the base application-record.
+
+An API controller with a good example of `create_resource` method is `auth_key_pairs_controller.rb`, so you can copy-paste from there and adjust as needed.
+
+If you are adding a new controller, you'll need to add routes to in `manageiq-api/api.yml`. Copy from one of the existing routes and adjust.
```

Comment:
If you're adding create to the collection you'll also need to edit `api.yml`, think it'd be `:verbs: *gpppd` but that always confuses me :laughing: assume it is `GET, PUT, PATCH, POST, DESTROY` but what if you only have `GET, POST` is it `*gp`?  How do we know that's not `GET, PATCH`? cc @Fryguy

Reply:
NOTE you also might need to create a new product feature

Reply:
yes, `gpppd` is `GET POST PUT PATCH DESTROY`.  The star at the beginning of `*gpppd` is a YAML anchor that points to the definition.  See https://github.com/ManageIQ/manageiq-api/blob/master/config/api.yml#L6-L31  We know it's not `GET PATCH`, because `GET PATCH` is not defined 😉

Reply:
:+1: so if you wanted to implement only GET and PATCH you could call them out and just not use the shorthand

Reply:
If it helps (or for others that aren't aware of YAML anchors)...

```ruby
[1] pry(main)> require "yaml"
=> true
[2] pry(main)> x = {:target => {:a => 1, :b => 2}}
=> {:target=>{:a=>1, :b=>2}}
[3] pry(main)> x[:reference] = x[:target]
=> {:a=>1, :b=>2}
[4] pry(main)> x[:reference].object_id == x[:target].object_id
=> true
[5] pry(main)> puts x.to_yaml
---
:target: &1
  :a: 1
  :b: 2
:reference: *1
=> nil
[6] pry(main)> y = YAML.load(x.to_yaml)
=> {:target=>{:a=>1, :b=>2}, :reference=>{:a=>1, :b=>2}}
[7] pry(main)> y[:reference].object_id == y[:target].object_id
=> true
```

Reply:
so we created this nomenclature to minimize typing??

Reply:
anchors are also used by Rails in `config/database.yml`, if I recall correctly

Reply:
It minimizes _a lot_ of duplication, but yes, it was created to declutter the file.  It also ensures that only known "sets" are chosen.

Reply:
For reference:

```ruby
[8] pry(main)> YAML.load_file("config/api.yml").deep_dup.to_yaml.lines.count
=> 5457
[9] pry(main)> File.read("config/api.yml").lines.count
=> 5068
```

The first line is fully expanded, the second line is the current, so we're talking 389 lines or 7.5% bigger

Reply:
@agrare, @Fryguy 
I added instructions about adding `create` to YAML. Could you see that it's correct and complete?

Also, could you help me with instructions for adding an appropriate product feature?